crypto/dh/dh_lib.c

```diff
@@ -246,14 +246,21 @@ void DH_get0_pqg(const DH *dh, BIGNUM **p, BIGNUM **q, BIGNUM **g)
 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 {
     /* q is optional */
-    if (p == NULL || g == NULL)
+    if ((p == NULL && dh->p == NULL)
+        || (g == NULL && dh->g == NULL))
```

Comment:
Side note: it dawned on me that the expression I suggested can be shortened like this:

``` C
    if ((p == NULL) == (dh->p == NULL)
        || (g == NULL) == (dh->g == NULL))
```

I wonder if other readers will understand the meaning of this form, though.  Basically, that implements logical equivalence.  Opinions?